### PR TITLE
Use crm_attribute --list-options

### PR DESCRIPTION
**Pull request by aleksei-burlakov**
Use crm_attribute --list-options
    
pacemaker-v3+ enables crm_attribute --list-options
 which is preferable over the "pacemaker-schedulerd",
"pacemaker-controld", "pacemaker-based", or even older
"pengine", "crmd", "cib".

### CODE REVIEW

**Comment by liangxin1300** [comment]
Do we have pacemaker 3.0 package available now?
If not, how to test this for now?

**Comment by zzhou1** [comment]
> Do we have pacemaker 3.0 package available now? If not, how to test this for now?

FYI, This aims for PED-8016 pacemaker 2.1.8 as well, which will be released almost the same time of pacemaker 3.0, in mid-2024. Not sure if @gao-yan has either of them somewhere?

**Comment by gao-yan** [comment]
A revision of pacemaker with "crm_attribute --list-options" is already in Factory.

**Comment by liangxin1300** [comment]
> A revision of pacemaker with "crm_attribute --list-options" is already in Factory.

Can you show me the link?
Thanks!

**Comment by aleksei-burlakov** [comment]
> > A revision of pacemaker with "crm_attribute --list-options" is already in Factory.
> 
> Can you show me the link? Thanks!

You can just use the tumbleweed, it's already there

File: hawk/app/models/crm_config.rb [hunk 1]
```diff
@@ -48,60 +48,71 @@ def persisted?
...
-                        value.strip
-                      end.reject do |value|
-                        value.empty?
-                      end
-
-                      crm_config[name][:values] = values unless values.empty?
-                    end
-                  end
-                end
+            # The crm_attribute is only available in pacemaker-v3+
```
Comment (liangxin1300) [inline hawk/app/models/crm_config.rb:98]:
```suggestion
            # The crm_attribute --list-options is only available since pacemaker 2.1.8
```

**Comment by liangxin1300** [comment]
It looks like the output of `/usr/libexec/pacemaker/pacemaker-fenced metadata` not included in crm_attribute --list-options

File: hawk/app/models/crm_config.rb (hunk 1 above)
Comment (aleksei-burlakov) [reply hawk/app/models/crm_config.rb:98]: Good catch!

**Comment by aleksei-burlakov** [comment]
> It looks like the output of `/usr/libexec/pacemaker/pacemaker-fenced metadata` not included in crm_attribute --list-options

That's right. For some reason there are no options from `pacemaker-attrd, pacemaker-execd, and pacemaker-fenced`

**Comment by gao-yan** [comment]
The metadata from `pacemaker-fenced metadata` is not about cluster options. It lists the supported parameters that can be configured for fencing resources.

The metadata is going to be provided from another CLI like  `crm_resource --list-options fencing --output-as xml`

**Comment by liangxin1300** [comment]
> The metadata is going to be provided from another CLI like `crm_resource --list-options fencing --output-as xml`

Since which version?

**Comment by gao-yan** [comment]
> > The metadata is going to be provided from another CLI like `crm_resource --list-options fencing --output-as xml`
> 
> Since which version?

It'll be in 2.1.8 and 3.0.0. And `pacemaker-fenced metadata` is deprecated but will be retained at least for 2.1.8 though.

BTW, there's a pacemaker revision from the main branch with the CLI "crm_resource --list-options fencing/primitive" submitted into the factory devel project.

File: hawk/app/models/crm_config.rb [hunk 2]
```diff
@@ -48,60 +48,71 @@ def persisted?
...
+          crm_config[name] = {
+            type: content.attributes["type"],
+            readonly: false,
+            shortdesc: shortdesc,
+            longdesc: longdesc,
+            advanced: (advanced and (advanced=="1")) ? true : false,
+            default: default
+          }
+
+          if type == "enum"
```
Comment (gao-yan) [inline hawk/app/models/crm_config.rb:76]: I'm confused where the type name "enum" is from :-)  It's called "select" type in OCF standard 1.1:

https://github.com/ClusterLabs/OCF-spec/blob/main/ra/1.1/ra-api.rng#L70

File: hawk/app/models/crm_config.rb [hunk 3]
```diff
@@ -48,60 +48,71 @@ def persisted?
...
+            match = longdesc.match(/Allowed values:(.*)/i)
```
Comment (gao-yan) [inline hawk/app/models/crm_config.rb:77]: And we should first try to retrieve the allowed values from the "option" elements under the "content" if appearing:

https://github.com/ClusterLabs/pacemaker/pull/3297/commits/5ef28e37d6f3008a55a7621262a5acd3be637c16#diff-14ca730a938353184deb85c9b9e22bc6e448635cfc7fbb62f8a80036ab97fe32R67-R70

Parsing from any "Allowed values ..." should be a fallback.

File: hawk/app/models/crm_config.rb (hunk 2 above)
Comment (aleksei-burlakov) [reply hawk/app/models/crm_config.rb:76]: I didn't change this part but simply moved this block of code. If it should be changed, I will create another PR.

File: hawk/app/models/crm_config.rb (hunk 3 above)
Comment (aleksei-burlakov) [reply hawk/app/models/crm_config.rb:77]: Good point. I didn't change this block but simply moved it. I will add retrieving the allowed values from the option as sub-feature in the next PR.

**Pull request merged**